consensus/service/src/byzantine_ledger.rs

```diff
@@ -76,16 +72,14 @@ pub struct ByzantineLedger {
 
 impl ByzantineLedger {
     pub fn new<
-        E: ConsensusEnclaveProxy,
         PC: BlockchainConnection + ConsensusConnection + 'static,
         L: Ledger + Sync + 'static,
-        UI: UntrustedInterfaces + Send + Sync + 'static,
     >(
         node_id: NodeID,
         quorum_set: QuorumSet,
         peer_manager: ConnectionManager<PC>,
         ledger: L,
-        tx_manager: TxManager<E, L, UI>,
+        tx_manager: Arc<Mutex<Box<dyn TxManager>>>,
```

Comment:
Would it make more sense to add a generic type parameter to `ByzantineLedger.new` for `TxManager` instead of using `Arc<Mutex<Box<dyn TxManager>>>`? This would mean the caller would lock the mutex and pass `&TxManagerGenericType` to `ByzantineLedger.new`, keeping the actual `ByzantineLedger` agnostic to the way `TxManager` is stored.

Reply:
Hmm. Yeah, we could add a type parameter ("TXM: TxManager") to ByzantineLedger, although I kind of like how this worked out with Box. Maybe it's my Java background, but I feel like this approach is closer to "dependency injection", which is roughly the idea I'm going for. Making ByzantineLedger generic w.r.t. TxManager could accomplish the same thing, but I feel like it bloats up ByzantineLedger's type signature unnecessarily.

Reply:
I was suggesting adding it to `ByzantineLedger.new` not `ByzantineLedger` itself.

This would mean any place with the generic type definition of
```
E: ConsensusEnclaveProxy,
UI: UntrustedInterfaces,
```
would become
```
TXM: TxManager,
```

Reply:
> but I feel like this approach is closer to "dependency injection", which is roughly the idea I'm going for.

well they are both dependency injection, its just that one is static and one is dynamic

> but I feel like it bloats up ByzantineLedger's type signature unnecessarily

i agree that the earlier form felt bad, because implementation details of tx manager became type parameters of byzantine ledger.
but the mutex is going to prevent parallel access to the tx manager, which was possible for the methods marked `&self` earlier. depending on the structure of the rest of the program, this can cause a deadlock

fwiw I think kyle's approach is good, and it gives a simple way to slightly back out of this PR without giving up on the mocking etc., and it also has fewer template parameters than before